src/Microsoft.AspNetCore.WebUtilities/FormReader.cs

```diff
@@ -198,19 +276,69 @@ private void Buffer()
         /// <param name="encoding">The character encoding to use.</param>
         /// <param name="cancellationToken">The <see cref="CancellationToken"/>.</param>
         /// <returns>The collection containing the parsed HTTP form body.</returns>
-        public static async Task<Dictionary<string, StringValues>> ReadFormAsync(Stream stream, Encoding encoding, CancellationToken cancellationToken = new CancellationToken())
+        public static ValueTask<Dictionary<string, StringValues>> ReadFormAsync(Stream stream, Encoding encoding, CancellationToken cancellationToken = default(CancellationToken))
         {
+            var isAsync = false;
             var reader = new FormReader(stream, encoding);
+            try
+            {
+                var accumulator = new KeyValueAccumulator();
 
-            var accumulator = new KeyValueAccumulator();
-            var pair = await reader.ReadNextPairAsync(cancellationToken);
-            while (pair.HasValue)
+                var pairTask = reader.ReadNextPairAsync(cancellationToken);
+                if (!pairTask.IsCompletedSuccessfully)
+                {
+                    isAsync = true;
+                    return ReadFormAsyncAwaited(pairTask.AsTask(), reader, accumulator, cancellationToken);
+                }
+
+                var pair = pairTask.Result;
+                while (pair.HasValue)
+                {
+                    accumulator.Append(pair.Value.Key, pair.Value.Value);
+
+                    pairTask = reader.ReadNextPairAsync(cancellationToken);
+                    if (!pairTask.IsCompletedSuccessfully)
+                    {
+                        isAsync = true;
+                        return ReadFormAsyncAwaited(pairTask.AsTask(), reader, accumulator, cancellationToken);
+                    }
+
+                    pair = pairTask.Result;
+                }
+
+                return accumulator.GetResults();
+            }
+            finally
             {
-                accumulator.Append(pair.Value.Key, pair.Value.Value);
-                pair = await reader.ReadNextPairAsync(cancellationToken);
+                if (!isAsync)
+                {
+                    reader.Dispose();
+                }
             }
+        }
 
-            return accumulator.GetResults();
+        private static async Task<Dictionary<string, StringValues>> ReadFormAsyncAwaited(Task<KeyValuePair<string, string>?> pairTask, FormReader reader, KeyValueAccumulator accumulator, CancellationToken cancellationToken)
+        {
+            using (reader)
+            {
+                var pair = await pairTask;
+                while (pair.HasValue)
```

Comment:
I imagine a golden time in the future when the compiler automatically does these patterns for you...